### PR TITLE
Basic auth package for authorising users

### PR DESCRIPTION
**Pull request by egnwd**
Currently this allows a service to be authorised using google OAuth. As it stands only the student side is wired. The current plan is to keep the sponsors and students separate, just incase auth changes later etc.

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+DB_NAME=sponsor_portal
+DB_HOST=localhost
+DB_PORT=5432
+DB_SSLMODE=disable
+PGPASSWORD=${DB_PASS}
+
+OAUTH_COOKIE_SECRET=topsecret
+OAUTH_BASE_URL=http://localhost:${PORT}
+OAUTH_CLIENT_ID=190477914499-dl3446c9h3t5k5dsidl9t1n1hnekhk3m.apps.googleusercontent.com
+OAUTH_CLIENT_SECRET=lWpdWF2JDHOBD7n4Q1WfyY9g
```
Comment (plietar) [inline .env]: Nice one

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.DS_STORE
```
Comment (plietar) [inline .gitignore]: This isn't really the place, but whatever
¯\\_(ツ)_/¯

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,29 @@
+include .env
+export $(shell sed 's/=.*//' .env)
+
+DB_CONN = postgres://${DB_USER}@${DB_HOST}:${DB_PORT}/${DB_NAME}?sslmode=${DB_SSLMODE}
```
Comment (plietar) [inline Makefile]: Use `:=` instead of `=`

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+.PHONY: start seed server client
+
+install:
+	glide install
+
+setupdb:
+	migrate -database ${DB_CONN} -path migrations up
+
+seed:
+	docker exec -i sponsorportal_sponsor_portal_db_1 sh -c "psql -U ${DB_USER} -d ${DB_NAME} -c \"COPY users (name, email) FROM STDIN WITH CSV\"" < seed/users.csv
```
Comment (plietar) [inline Makefile]: Break that into multiple lines

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,29 @@
+include .env
+export $(shell sed 's/=.*//' .env)
```
Comment (plietar) [inline Makefile]: Is this necessary ? Not great to be leaking credentials into the environment

File: Makefile (hunk 4 above)
Comment (plietar) [reply Makefile]: Also the container name is very specific

File: auth/middleware.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"context"
+	"net/http"
+)
+
+type userKeyType int
+
+const userKey userKeyType = iota
+
+func (auth *Auth) RequireAuth(inner http.HandlerFunc) http.HandlerFunc {
```
Comment (plietar) [inline auth/middleware.go]: `inner` should be a `http.Handler`. That would be more general and reusable than an `HandlerFunc`

File: auth/middleware.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		userId, err := auth.getCurrentUser(r)
+		if err != nil {
+			http.Error(w, "Failed to get current user", http.StatusInternalServerError)
```
Comment (plietar) [inline auth/middleware.go]: The error should be logged somewhere

File: auth/middleware.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+			return
+		}
+
+		rWithUser := SetRequestUser(r, userId)
+
+		inner(w, rWithUser)
+	})
+}
+
+func SetRequestUser(r *http.Request, userId UserIdentifier) *http.Request {
```
Comment (plietar) [inline auth/middleware.go]: Shouldn't that be private ?

File: auth/user.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+	err = session.Save(r, w)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (auth *Auth) User(r *http.Request) UserIdentifier {
```
Comment (plietar) [inline auth/user.go]: What's the difference between this function and the one in `middleware.go` ?

Comment (egnwd) [inline auth/user.go]: _@plietar clicked the wrong thing and deleted your comment, meant to delete mine..._

You are correct, will remove one

File: Makefile (hunk 4 above)
Comment (egnwd) [reply Makefile]: Yep, I think I need some sort of `$(docker inspect ...)` will look into that

File: auth/user.go (hunk 9 above)
Comment (plietar) [reply auth/user.go]: A N G E R Y

File: auth/callback.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+func (auth *Auth) handleCallback(w http.ResponseWriter, r *http.Request) {
+	expectedState, err := auth.getAndDeleteState(w, r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	state := r.FormValue("state")
+	if state != expectedState {
+		http.Error(w, fmt.Sprintf("invalid oauth state, expected '%s', got '%s'", expectedState, state), http.StatusInternalServerError)
```
Comment (martinzlocha) [inline auth/callback.go]: Line too long. Break up into two.

File: auth/callback.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+	"io/ioutil"
+	"net/http"
+
+	"golang.org/x/oauth2"
+)
+
+func (auth *Auth) handleCallback(w http.ResponseWriter, r *http.Request) {
+	expectedState, err := auth.getAndDeleteState(w, r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (martinzlocha) [inline auth/callback.go]: Does Node by default log the errors somewhere? It would be good to store them somewhere.

File: .gitignore (hunk 2 above)
Comment (egnwd) [reply .gitignore]: Removed it anyway...

File: auth/callback.go (hunk 11 above)
Comment (egnwd) [reply auth/callback.go]: I add a simple log line above, but I think we need a separate PR for proper logging

**Comment by plietar** [comment]
README needs instructions on setting up the database

**Review by martinzlocha — APPROVED** [review]

**Pull request merged**